### PR TITLE
Manual migration CLI command

### PR DESCRIPTION
**Pull request by KPrasch**
**Type of PR:**
Feature

**Required reviews:** 
2

**What this does:**
Provides a CLI command for manual period migration.

**Why it's needed:**
Provides an action for stakers who have unbonded stakes during period migration who will otherwise be unable to make commitments.

### CODE REVIEW

File: nucypher/cli/commands/stake.py [hunk 1]
```diff
@@ -1349,3 +1349,43 @@ def mint(general_config: GroupGeneralConfig,
...
+    STAKEHOLDER = transacting_staker_options.create_character(emitter, config_file)
+    blockchain = transacting_staker_options.get_blockchain()
+    client_account, staking_address = select_client_account_for_staking(
+        emitter=emitter,
+        stakeholder=STAKEHOLDER,
+        staking_address=transacting_staker_options.staker_options.staking_address)
+
+    # Interact
+    if STAKEHOLDER.staker.is_migrated:
+        emitter.echo('Saker has already migrated.', color='red')
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1374]:
```suggestion
        emitter.echo('Staker has already migrated.', color='red')
```

File: nucypher/cli/commands/stake.py [hunk 2]
```diff
@@ -1349,3 +1349,43 @@ def mint(general_config: GroupGeneralConfig,
...
+        raise click.Abort
+    if not force:
+        click.confirm('Confirm manual migration', abort=True)
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1377]: Should we move to literature?

**Review by derekpierre — COMMENTED** [review]
@KPrasch  / @vzotova to make our lives easier with support, wdyt about adding a line entry about the value of the migrated flag to the output of `nucypher stake list` (https://github.com/nucypher/nucypher/blob/main/nucypher/cli/painting/staking.py#L80) and `nucypher status stakers` (https://github.com/nucypher/nucypher/blob/main/nucypher/cli/painting/status.py#L190).  However, I think we should print the extra line only if migrated is false; if migrated is true we shouldn't print anything extra since this should be the norm after the transition period and it would be redundant to print it as true.

**Comment by derekpierre** [comment]
Also, we should add information to the CLI docs about the additional `migrate` command.

File: nucypher/cli/commands/stake.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py]: ✔️

File: nucypher/cli/commands/stake.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py]: ✔️

File: nucypher/cli/commands/stake.py [hunk 3]
```diff
@@ -1349,3 +1351,48 @@ def mint(general_config: GroupGeneralConfig,
...
                           transaction_type='mint')
+
+
+@stake.command()
+@group_transacting_staker_options
+@option_config_file
+@option_force
+@group_general_config
+@click.option('--external-staker-address',
+              help="The staking address of a staker to manually migrate", type=EIP55_CHECKSUM_ADDRESS)
```
Comment (cygnusv) [inline nucypher/cli/commands/stake.py:1362]: What's the rationale for this argument?

Comment (vzotova) [reply nucypher/cli/commands/stake.py:1362]: Will be helpful for pool or if when will remain a few not migrated stakers - we can do this manually and then remove migration code from contract

File: nucypher/blockchain/eth/agents.py [hunk 4]
```diff
@@ -601,6 +601,20 @@ def set_restaking(self, transacting_power: TransactingPower, value: bool) -> TxR
         # TODO: Handle ReStakeSet event (see #1193)
         return receipt
 
+    @contract_api(CONTRACT_CALL)
+    def is_migrated(self, staker_address: ChecksumAddress) -> bool:
+        flags = self.get_flags(staker_address)
+        return flags.migration_flag
+
+    @contract_api(TRANSACTION)
+    def migrate(self, transacting_power: TransactingPower, staker_address: ChecksumAddress = None) -> TxReceipt:
```
Comment (derekpierre) [inline nucypher/blockchain/eth/agents.py:610]: Should `staker_address` be Optional?

```suggestion
    def migrate(self, transacting_power: TransactingPower, staker_address: Optional[ChecksumAddress] = None) -> TxReceipt:
```

File: nucypher/cli/commands/stake.py [hunk 5]
```diff
@@ -1349,3 +1351,48 @@ def mint(general_config: GroupGeneralConfig,
...
+
+    # Authenticate
+    password = get_password(stakeholder=STAKEHOLDER,
+                            blockchain=blockchain,
+                            client_account=client_account,
+                            hw_wallet=transacting_staker_options.hw_wallet)
+    STAKEHOLDER.assimilate(checksum_address=client_account, password=password)
+
+    # Migrate
+    receipt = STAKEHOLDER.staker.migrate(staker_address=staking_address)
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1394]: Should this be `address_to_migrate` 🐞 ?

```suggestion
    receipt = STAKEHOLDER.staker.migrate(staker_address=address_to_migrate)
```

File: nucypher/cli/commands/stake.py (hunk 3 above)
Comment (derekpierre) [reply nucypher/cli/commands/stake.py:1362]: How about something like `--staker-to-migrate`...? We should also clarify that if not specified, the default is the staking address used for transacting.

File: nucypher/cli/commands/stake.py [hunk 6]
```diff
@@ -1349,3 +1351,48 @@ def mint(general_config: GroupGeneralConfig,
...
+        stakeholder=STAKEHOLDER,
+        staking_address=transacting_staker_options.staker_options.staking_address)
+
+    # Interact
+    address_to_migrate = external_staker_address or staking_address
+    if STAKEHOLDER.staker.is_migrated:
+        emitter.echo(MIGRATION_ALREADY_PERFORMED.format(address=address_to_migrate), color='red')
+        raise click.Abort
+    if not force:
+        click.confirm(CONFIRM_MANUAL_MIGRATION.format(address=address_to_migrate), abort=True)
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1384]: 👍

File: nucypher/cli/commands/stake.py [hunk 7]
```diff
@@ -1349,3 +1351,48 @@ def mint(general_config: GroupGeneralConfig,
...
+
+
+@stake.command()
+@group_transacting_staker_options
+@option_config_file
+@option_force
+@group_general_config
+@click.option('--external-staker-address',
+              help="The staking address of a staker to manually migrate", type=EIP55_CHECKSUM_ADDRESS)
+def migrate(general_config: GroupGeneralConfig,
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1363]: We should add a docstring so that it is used as the help text for the CLI

**Review by derekpierre — CHANGES_REQUESTED** [review]

File: nucypher/cli/commands/stake.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py:1394]: Yes, you're right! Thank you.

File: nucypher/cli/commands/stake.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py:1362]: Meh, Comes off a bit redundant to me: `nucypher stake migrate --staker-to-migrate`.

File: nucypher/blockchain/eth/agents.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/agents.py]: ✔️

File: nucypher/cli/commands/stake.py (hunk 7 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py:1363]: ✔️

File: nucypher/cli/commands/stake.py (hunk 5 above)
Comment (KPrasch) [reply nucypher/cli/commands/stake.py]: ✔️

File: nucypher/cli/commands/stake.py [hunk 8]
```diff
@@ -1349,3 +1351,52 @@ def mint(general_config: GroupGeneralConfig,
...
+@click.option('--external-staker-address',
+              help="The staking address to manually migrate (defaults to wallet address)", type=EIP55_CHECKSUM_ADDRESS)
+def migrate(general_config: GroupGeneralConfig,
+            transacting_staker_options: TransactingStakerOptions,
+            config_file,
+            force,
+            external_staker_address):
+    """
+    Manually migrate a staker to updated period duration. Stakers do not normally need to run
+    this command unless there is a problem with automated migration or commitments..
```
Comment (derekpierre) [inline nucypher/cli/commands/stake.py:1370]:
```suggestion
    this command unless there is a problem with automated migration or commitments.
```

**Review by derekpierre — APPROVED** [review]
🎸

**Review by cygnusv — APPROVED** [review]

**Comment by derekpierre** [comment]
@KPrasch do you want to get https://github.com/nucypher/nucypher/pull/2609/commits/712adb0be07b5df4f0073f0226e9264cb3945910 into `main` directly, or are we good to wait for `dao-propsoal-1` EPIC to get merged into `main` (presumably today)?

Also, eventually the CLI test will fail again as we get closer to the end of probationary period (April 30th). We could think about a more seamless fix that will continue to work closer to the end of the probationary period, and after the probationary period is over.

**Comment by KPrasch** [comment]
@derekpierre 

> do you want to get 712adb0 into main directly, or are we good to wait for dao-propsoal-1 EPIC to get merged into main (presumably today)?

It will make it's way into main later today with a new release anyways.

> Also, eventually the CLI test will fail again as we get closer to the end of probationary period (April 30th). We could think about a more seamless fix that will continue to work closer to the end of the probationary period, and after the probationary period is over - potentially some override for tests...?

Agreed!

**Pull request merged**